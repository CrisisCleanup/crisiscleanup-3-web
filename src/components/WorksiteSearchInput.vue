<template>
  <div class="flex items-center justify-start w-full autocomplete">
    <vue-autosuggest
      v-bind="$attrs"
      :suggestions="suggestions"
      :input-props="{
        id: 'autosuggest__input',
        placeholder,
        required,
        autocomplete: 'off',
        type: 'search',
      }"
      :section-configs="sectionConfigs"
      :render-suggestion="renderSuggestion"
      :get-suggestion-value="getSuggestionValue"
      :should-render-suggestions="shouldRenderSuggestions"
      :class="classes"
      @input="onInputChange"
    />
    <div
      v-if="loading"
      class="icon-container flex items-center justify-center"
      :class="iconClasses"
    >
      <font-awesome-icon icon="spinner" spin />
    </div>
    <div
      v-if="(icon || tooltip) && !loading"
      class="icon-container flex items-center justify-center"
      :class="iconClasses"
    >
      <ccu-icon :type="tooltip ? 'info' : icon" size="small" />
    </div>
  </div>
</template>

<script>
import Highlighter from 'vue-highlight-words';
import Worksite from '@/models/Worksite';
import User from '@/models/User';
import { getWorkTypeImage } from '@/filters';

export default {
  name: 'WorksiteSearchInput',
  props: {
    suggestions: {
      type: Array,
      default: () => [],
    },
    displayProperty: {
      type: String,
      default: '',
    },
    icon: {
      type: String,
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: Boolean,
    size: {
      type: String,
      default: '',
    },
    tooltip: {
      type: String,
      default: '',
    },
    full: Boolean,
    loading: Boolean,
    width: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      selected: '',
      filteredOptions: [],
      timeout: null,
      debounceMilliseconds: 50,
      inputProps: {
        id: 'autosuggest__input',
        onInputChange: this.onInputChange,
        placeholder: "Type 'e'",
      },
      classes: {
        'flex-grow': true,
        relative: true,
        'text-base': true,
        'font-light': true,
        large: this.size === 'large',
        base: this.size !== 'large',
        'has-icon': Boolean(this.icon),
        'has-tooltip': Boolean(this.tooltip),
        full: Boolean(this.full),
      },
      iconClasses: {
        large: this.size === 'large',
        base: this.size !== 'large',
        'has-tooltip': Boolean(this.tooltip),
      },
      text: '',
      sectionConfigs: {
        worksites: {
          limit: 5,
          label: this.$t('Existing Cases'),
          type: '',
          onSelected: selected => {
            this.selected = selected.item;
            this.$emit('selectedExisting', selected.item);
          },
        },
        geocoder: {
          limit: 10,
          label: this.$t('Geocoder'),
          type: '',
          onSelected: selected => {
            this.selected = selected.item;
            this.$emit('selectedGeocode', selected.item);
          },
        },
      },
    };
  },
  computed: {
    currentUser() {
      return User.find(this.$store.getters['auth/userId']);
    },
  },
  methods: {
    shouldRenderSuggestions(size, loading) {
      return size > 0 && !loading;
    },
    renderSuggestion(suggestion) {
      if (suggestion.name === 'geocoder') {
        return (
          <div class="flex flex-col text-sm p-1 cursor-pointer hover:bg-gray-100 border-b">
            <div>{suggestion.item.description}</div>
          </div>
        );
      }
      return (
        <div class="flex items-center p-1 cursor-pointer hover:bg-gray-100 border-b">
          <span>{this.getWorkImage(suggestion.item.work_types)}</span>
          <div
            className="flex flex-col text-sm"
            style={{ width: this.width || 'auto' }}
          >
            <Highlighter
              highlightClassName="highlight"
              searchWords={[this.text]}
              autoEscape={true}
              textToHighlight={`${suggestion.item.name}, ${suggestion.item.case_number}`}
            />
            <br />
            <Highlighter
              highlightClassName="highlight"
              searchWords={[this.text]}
              autoEscape={true}
              textToHighlight={`${suggestion.item.address}, ${suggestion.item.city}, ${suggestion.item.state}`}
            />
          </div>
        </div>
      );
    },
    getSuggestionValue(suggestion) {
      if (suggestion.name === 'geocoder') {
        return suggestion.item.description;
      }
      return suggestion.item.name;
    },
    getWorkImage(workTypes) {
      const workType = Worksite.getWorkType(
        workTypes,
        null,
        this.currentUser.organization,
      );

      if (!workType) {
        return '';
      }

      const svg = getWorkTypeImage(workType);

      return this.$createElement('div', {
        domProps: {
          innerHTML: svg,
        },
        class: {
          'case-svg-container': true,
          'mr-1': true,
        },
      });
    },
    onInputChange(text) {
      clearTimeout(this.timeout);
      this.timeout = setTimeout(() => {
        if (text === '' || text === undefined) {
          return;
        }
        this.$emit('search', text);
        this.$emit('input', text);
        this.text = text;
      }, this.debounceMilliseconds);
    },
  },
};
</script>

<style>
.case-svg-container svg {
  width: 30px;
  height: 30px;
}

#autosuggest__input {
  outline: none;
  width: 100%;
  height: 32px;
  border-radius: 0;
  border: solid 1px #dadada;
  padding: 10px;
  position: relative;
}

.large #autosuggest__input {
  height: 50px;
}

.has-icon #autosuggest__input {
  width: 100%;
  border-right: 0;
}

.has-tooltip #autosuggest__input {
  width: 100%;
}

.autosuggest__results {
  font-weight: 300;
  min-width: 120%;
  max-width: 100vw;
  position: absolute;
  z-index: 10000001;
  border: 1px solid #e0e0e0;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  background: white;
  padding: 0;
  overflow: auto;
  max-height: 200px;
}

.full .autosuggest__results {
  width: max-content;
}

.autocomplete .icon-container {
  width: 32px;
  height: 32px;
  border: solid 1px #dadada;
  border-left: 0;
}

.autocomplete .has-tooltip.icon-container {
  background-color: #f7f7f7;
}

.autocomplete .large.icon-container {
  width: 50px;
  height: 50px;
}

.autosuggest__results-before {
  @apply text-gray-400 text-sm font-bold px-1;
}
</style>
