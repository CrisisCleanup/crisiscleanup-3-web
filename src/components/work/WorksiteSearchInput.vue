<template>
  <div class="flex items-center justify-start autocomplete">
    <div class="relative w-full">
      <base-input
        :model-value="value"
        :size="size"
        :placeholder="placeholder"
        :required="required"
        :tooltip="tooltip"
        @update:modelValue="worksitesSearch"
        @input.stop=""
        @focus="isFocused = true"
        @blur="onBlur"
      >
      </base-input>

      <div
        v-if="results.length > 0 && value.length > 0 && isFocused"
        class="absolute bg-white z-50 h-auto max-h-84 overflow-auto min-w-84"
      >
        <div v-for="result in results" :key="result.label">
          <template v-if="result.options.length > 0"
            >{{ result.label }}
            <div v-for="option in result.options" :key="option">
              <div
                v-if="result.label === 'Geocode'"
                class="flex flex-col sm:text-lg text-base p-1 cursor-pointer hover:bg-crisiscleanup-light-grey border-b"
                @click="() => $emit('selectedGeocode', option)"
              >
                <div>{{ option.description }}</div>
              </div>

              <div
                v-else
                class="flex items-center p-1 cursor-pointer hover:bg-crisiscleanup-light-grey border-b"
                @click="() => $emit('selectedExisting', option)"
              >
                <div
                  class="mr-1 case-svg-container"
                  v-html="getWorkImage(option.work_types)"
                ></div>
                <div
                  class="flex flex-col text-sm"
                  :style="{ width: width || 'auto' }"
                >
                  {{ option.name }}, {{ option.case_number }}
                  <br />
                  {{ option.address }}, {{ option.city }}, {{ option.state }}
                </div>
              </div>
            </div>
          </template>
        </div>
      </div>
    </div>

    <div
      v-if="icon || tooltip"
      class="icon-container flex items-center justify-center"
      :class="iconClasses"
    >
      <ccu-icon
        :alt="$t('actions.help_alt')"
        :type="tooltip ? 'info' : icon"
        size="small"
      />
    </div>
  </div>
</template>

<script>
// import Highlighter from 'vue-highlight-words';
// import Worksite from '../../models/Worksite';
// import { getWorkTypeImage } from '../../filters';
import { computed, nextTick, ref } from 'vue';
import axios from 'axios';
import { useStore } from 'vuex';
import useCurrentUser from '../../hooks/useCurrentUser';
import GeocoderService from '../../services/geocoder.service';
import BaseInput from '../BaseInput.vue';
import Worksite from '../../models/Worksite';
import { getWorkTypeImage } from '../../filters/index';

export default {
  name: 'WorksiteSearchInput',
  components: { BaseInput },
  props: {
    value: {
      type: String,
      default: '',
    },
    icon: {
      type: String,
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: Boolean,
    size: {
      type: String,
      default: '',
    },
    tooltip: {
      type: String,
      default: '',
    },
    full: Boolean,
    width: {
      type: String,
      default: '',
    },
    selector: {
      type: String,
      default: '',
    },
    skipValidation: {
      type: Boolean,
    },
    useGeocoder: {
      type: Boolean,
    },
    useWorksites: {
      type: Boolean,
      default: true,
    },
  },
  setup(props, { emit }) {
    const { currentUser } = useCurrentUser();
    const store = useStore();
    const currentIncidentId = computed(
      () => store.getters['incident/currentIncidentId'],
    );
    window.addEventListener('keydown', (e) => {
      if (e.key === 'Escape' || e.key === 'Tab') {
        emit('clearSuggestions');
      }
    });

    function searchWorksites(search, incident) {
      return axios.get(
        `${
          import.meta.env.VITE_APP_API_BASE_URL
        }/worksites?fields=id,name,address,case_number,postal_code,city,state,incident,work_types&limit=5&search=${search}&incident=${incident}`,
      );
    }

    async function geocoderSearch(value) {
      return await GeocoderService.getMatchingAddresses(value, 'USA');
    }

    const results = ref([]);

    function onBlur() {
      setTimeout(() => {
        isFocused.value = false;
      }, 200);
    }

    async function worksitesSearch(value) {
      emit('input', value);
      let geocode = [];
      let worksites = [];
      if (props.useWorksites) {
        const sites = await searchWorksites(value, currentIncidentId.value);
        worksites = sites.data.results;
      }
      if (props.useGeocoder) {
        geocode = await geocoderSearch(value);
      }
      results.value = [
        {
          label: 'Search',
          options: worksites,
        },
        {
          label: 'Geocode',
          options: geocode,
        },
      ];
    }

    const selected = ref('');
    const isFocused = ref(false);
    const filteredOptions = ref([]);
    const timeout = ref(null);
    const debounceMilliseconds = ref(50);
    const classes = ref({
      'flex-grow': true,
      relative: true,
      'text-base': true,
      'font-light': true,
      large: props.size === 'large',
      base: props.size !== 'large',
      'has-icon': Boolean(props.icon),
      'has-tooltip': Boolean(props.tooltip),
      full: Boolean(props.full),
      invalid: !props.skipValidation,
    });
    const iconClasses = ref({
      large: props.size === 'large',
      base: props.size !== 'large',
      'has-tooltip': Boolean(props.tooltip),
    });
    const text = ref('');

    // shouldRenderSuggestions(size, loading) {
    //   return size > 0 && !loading;
    // },
    // renderSuggestion(suggestion) {
    //   if (suggestion.name === 'geocoder') {
    //     return (
    //       <div class="flex flex-col sm:text-lg text-base p-1 cursor-pointer hover:bg-crisiscleanup-light-grey border-b">
    //         <div>{suggestion.item.description}</div>
    //       </div>
    //     );
    //   }
    //   return (
    //     <div class="flex items-center p-1 cursor-pointer hover:bg-crisiscleanup-light-grey border-b">
    //       <span>{this.getWorkImage(suggestion.item.work_types)}</span>
    //       <div
    //         className="flex flex-col text-sm"
    //         style={{ width: this.width || 'auto' }}
    //       >
    //         <Highlighter
    //           highlightClassName="highlight"
    //           searchWords={[this.text]}
    //           autoEscape={true}
    //           textToHighlight={`${suggestion.item.name}, ${suggestion.item.case_number}`}
    //         />
    //         <br />
    //         <Highlighter
    //           highlightClassName="highlight"
    //           searchWords={[this.text]}
    //           autoEscape={true}
    //           textToHighlight={`${suggestion.item.address}, ${suggestion.item.city}, ${suggestion.item.state}`}
    //         />
    //       </div>
    //     </div>
    //   );
    // },
    // getSuggestionValue(suggestion) {
    //   if (suggestion.name === 'geocoder') {
    //     return suggestion.item.description;
    //   }
    //   return suggestion.item.name;
    // },
    function getWorkImage(workTypes) {
      const workType = Worksite.getWorkType(
        workTypes,
        null,
        currentUser.organization,
      );

      if (!workType) {
        return '';
      }

      return getWorkTypeImage(workType);
    }

    return {
      selected,
      filteredOptions,
      timeout,
      debounceMilliseconds,
      classes,
      iconClasses,
      text,
      currentUser,
      worksitesSearch,
      results,
      getWorkImage,
      isFocused,
      onBlur,
    };
  },
};
</script>

<style>
.autocomplete {
  z-index: 1000;
}

.case-svg-container svg {
  width: 30px;
  height: 30px;
}

#autosuggest__input {
  outline: none;
  width: 100%;
  height: 32px;
  border-radius: 0;
  border: solid 1px #dadada;
  padding: 10px;
  position: relative;
  @apply text-sm;
}

.invalid #autosuggest__input[required] {
  @apply border-crisiscleanup-red-100;
}

.large #autosuggest__input {
  height: 50px;
  @apply text-base;
}

.has-icon #autosuggest__input {
  width: 100%;
  border-right: 0;
}

.has-tooltip #autosuggest__input {
  width: 100%;
}

#autosuggest__input::placeholder {
  @apply text-crisiscleanup-dark-200;
}

#autosuggest__input {
  -webkit-appearance: none;
}

.autosuggest__results {
  font-weight: 200;
  min-width: 10vw;
  max-width: 100vw;
  position: absolute;
  z-index: 10000001;
  border: 1px solid #e0e0e0;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  background: white;
  padding: 0;
  overflow: auto;
  max-height: 400px;
}

.full .autosuggest__results {
  width: max-content;
}

.autocomplete .icon-container {
  width: 40px;
  height: 40px;
  border: solid 1px #dadada;
  border-left: 0;
}

.autocomplete .has-tooltip.icon-container {
  background-color: #f7f7f7;
}

.autocomplete .large.icon-container {
  width: 50px;
  height: 50px;
}

.autosuggest__results-before {
  @apply text-gray-400 text-sm font-bold px-1;
}

@media screen and (max-width: 640px) {
  .autosuggest__results {
    min-width: 97vw;
  }
}
</style>
