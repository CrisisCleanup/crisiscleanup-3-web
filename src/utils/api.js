const ROOT = process.env.VUE_APP_API_BASE_URL;

const getApiUrl = (path) => {
  const rootUrl = new URL(path, ROOT).toString();
  const getUrl = (endpoint) => {
    if (endpoint === undefined) {
      return rootUrl.slice(0, rootUrl.length - 1);
    }
    const fullUrl = new URL(endpoint, rootUrl).toString();
    return fullUrl;
  };
  return getUrl;
};

export const AgentApi = (endpoint) => getApiUrl('agents/')(endpoint);
export const PhoneApi = (endpoint) => getApiUrl('phone/')(endpoint);
export const IDPApi = (endpoint) => getApiUrl('idp/')(endpoint);
