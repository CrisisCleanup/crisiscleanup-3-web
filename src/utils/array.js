export function groupBy(array, key) {
  const result = {};
  array.forEach((item) => {
    if (!result[item[key]]) {
      result[item[key]] = [];
    }
    result[item[key]].push(item);
  });
  return result;
}
