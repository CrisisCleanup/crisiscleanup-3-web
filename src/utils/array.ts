export function groupBy(array, key) {
  const result = {};
  for (const item of array) {
    if (!result[item[key]]) {
      result[item[key]] = [];
    }

    result[item[key]].push(item);
  }

  return result;
}

export function childrenBy(dict, key) {
  const array = dict.null;
  for (const item of array) {
    item.children = dict[item[key]];
  }

  return array;
}
