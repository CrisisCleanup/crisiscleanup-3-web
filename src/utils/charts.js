let _seed = Date.now();
export function rand(min, max) {
  const seed = _seed;
  min = min === undefined ? 0 : min;
  max = max === undefined ? 1 : max;
  _seed = (seed * 9301 + 49297) % 233280;
  return min + (_seed / 233280) * (max - min);
}
