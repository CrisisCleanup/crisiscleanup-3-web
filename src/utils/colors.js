export const colors = {
    'open_unassigned': {
        color: '#d0021b',
    },
    'open_assigned': {
        color: '#d0021b',
    },
    'open_partially-completed': {
        color: '#d0021b',
    },
    'open_needs-follow-up': {
        color: '#d0021b',
    },
    'open_unresponsive': {
        color: '#787878',
    },
    'closed_completed': {
        color: '#82d78c',
    },
    'closed_partially-completed': {
        color: '#82d78c',
    },
    'closed_incomplete': {
        color: '#1d1d1d',
    },
    'closed_out-of-scope': {
        color: '#787878',
    },
    'closed_done-by-others': {
        color: '#0fa355',
    },
    'closed_no-help-wanted': {
        color: '#0fa355',
    },
    'closed_rejected': {
        color: '#0fa355',
    },
    'closed_duplicate': {
        color: '#0fa355',
    },
    'closed_marked-for-deletion': {
        color: '#0fa355',
    },
};