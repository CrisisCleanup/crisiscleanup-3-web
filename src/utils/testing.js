/**
 *
 * Testing Helpers
 *
 * utils
 */

export const MockOrganizations = [
  {
    id: 1,
    name: 'Test Organization',
    url: null,
    is_active: null,
    org_verified: null,
    created_at: '2019-12-25T18:24:05Z',
    updated_at: '2019-12-25T18:24:05Z',
  },
];

export const MockUsers = [
  {
    id: 1,
    first_name: 'First',
    last_name: 'Last',
    organization: {
      id: 1,
      name: 'Test Organization',
    },
  },
  {
    id: 2,
    first_name: 'Test',
    last_name: 'Last',
    organization: {
      id: 1,
      name: 'Test Organization',
    },
  },
];

export const MockWorkTypes = [
  {
    id: 1,
    name_t: 'workType.ash',
    description_t: 'workTypeDescription.ash',
    key: 'ash',
    file_prefix: 'Ash',
    commercial_value: 15000,
    depreciated_at: null,
    phase: null,
  },
  {
    id: 2,
    name_t: 'workType.biohazard',
    description_t: 'workTypeDescription.biohazard',
    key: 'biohazard',
    file_prefix: 'Biohazard',
    commercial_value: 50000,
    depreciated_at: null,
    phase: null,
  },
  {
    id: 7,
    name_t: 'workType.deferred_maintenance',
    description_t: 'workTypeDescription.deferred_maintenance',
    key: 'deferred_maintenance',
    file_prefix: 'Deferred_Maintenance',
    commercial_value: 10000,
    depreciated_at: null,
    phase: null,
  },
  {
    id: 16,
    name_t: 'workType.flood_indoor',
    description_t: 'workTypeDescription.flood_indoor',
    key: 'flood_indoor',
    file_prefix: 'Flood_indoor',
    commercial_value: 18000,
    depreciated_at: null,
    phase: null,
  },
  {
    id: 29,
    name_t: 'workType.pda_1',
    description_t: 'workTypeDescription.pda_1',
    key: 'pda_1',
    file_prefix: 'pda_1',
    commercial_value: 0,
    depreciated_at: null,
    phase: null,
  },
];

export const MockWorksites = [
  {
    id: 1,
    case_number: 'A196',
    name: 'First Last',
    updated_at: '2019-12-25T18:25:21.467128Z',
    location: {
      type: 'Point',
      coordinates: [-82.0535959749408, 29.8670584991275],
    },
    work_types: [
      {
        work_type: 'flood_indoor',
        status: 'closed_incomplete',
        claimed_by: null,
      },
      {
        work_type: 'pda_1',
        status: 'closed_incomplete',
        claimed_by: null,
      },
    ],
  },
  {
    id: 2,
    case_number: 'A68',
    name: 'Test Last',
    updated_at: '2019-12-25T18:25:21.494727Z',
    location: {
      type: 'Point',
      coordinates: [-82.3116865679154, 29.9818650922746],
    },
    work_types: [
      {
        work_type: 'pda_5',
        status: 'open_partially-completed',
        claimed_by: null,
      },

      {
        work_type: 'deferred_maintenance',
        status: 'closed_no-help-wanted',
        claimed_by: null,
      },
      {
        work_type: 'pda_1',
        status: 'closed_completed',
        claimed_by: null,
      },
    ],
  },
];

export const MockWorksiteRequest = {};

export const MockIncident = {
  id: 99,
  name: 'Texas Flooding, Oct 2018',
  case_label: 'A',
  start_at: '2018-10-17T00:00:00Z',
  form_fields: [
    {
      label_t: 'formLabels.phone1',
      html_type: 'text',
      data_sensitivity: 'verySensitive',
      data_group: 'personalInfo',
      help_t: 'formLabels.phone1_help',
      placeholder_t: 'formLabels.phone1_placeholder',
      is_required_default: true,
      is_read_only_default: false,
      read_only_break_glass: false,
      values_default: null,
      validation: 'string',
      values: null,
      is_required: null,
      is_read_only: null,
      list_order: 11,
      replaced_at: null,
      field_key: 'phone1',
      field_parent_key: 'property_info',
      if_selected_then_work_type: null,
      phase: 4,
    },
  ],
};

export const MockPhoneStatus = [
  {
    id: 1,
    substatus_name_t: 'Added work request to Crisis Cleanup',
    status_name_t: 'Answered: Added work request to Crisis Cleanup',
    primary_state_name_t: 'Answered',
    primary_state: 'answered',
    status: 'answered_added',
    substatus: 'added',
    completion: 1.0,
    try_again_delay: 0,
    list_order: 50,
    created_at: '2018-10-29T04:57:13Z',
    updated_at: '2020-03-25T04:19:27Z',
    must_associate_worksite: 'never',
    created_by: 14,
    updated_by: 14,
  },
];

export const MockAgents = [
  {
    agent_id: 'xxxx',
    instance_id: 'instance',
    arn: 'agentarn',
    user: MockUsers[0],
  },
];

export const mockModel = (data, { api, statics } = {}) => ({
  create: jest.fn(() => data[0]),
  all: jest.fn(() => data),
  find: jest.fn(() => data[0]),
  database: jest.fn(() => data),
  query: () => ({
    where: () => ({
      get: () => data[0],
    }),
    ...statics,
  }),
  api: jest.fn(() => ({
    get: () => data[0],
    fetch: () => data[0],
    ...api,
  })),
});
