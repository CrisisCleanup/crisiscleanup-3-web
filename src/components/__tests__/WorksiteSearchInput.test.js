/**
 *
 * Tests for WorksiteSearchInput
 *
 * Components
 */

import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import { MockWorksites } from '@/utils/testing';
import WorksiteSearchInput from '../WorksiteSearchInput';

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock('@/models/Worksite');
jest.mock('@/models/User');

const mocks = {
  $t: (key) => key,
};

const mountWithOptions = (store) =>
  shallowMount(WorksiteSearchInput, {
    stubs: ['vue-autosuggest', 'font-awesome-icon'],
    propsData: {
      suggestions: [{ name: 'worksites', data: [], key: 'name' }],
    },
    localVue,
    store,
    mocks,
  });

describe('WorksiteSearchInput', () => {
  let store;

  beforeEach(() => {
    store = new Vuex.Store();
  });
  it('should not log any errors', () => {
    const spy = jest.spyOn(global.console, 'error');
    mountWithOptions(store);
    expect(spy).not.toHaveBeenCalled();
  });

  it('should retrieve current user', () => {
    const wrapper = mountWithOptions(store);
    expect(wrapper.vm.currentUser).toMatchSnapshot();
  });

  it('should determine if it should render suggestions', () => {
    const wrapper = mountWithOptions(store);
    expect(wrapper.vm.shouldRenderSuggestions(0, false)).toBe(false);
    expect(wrapper.vm.shouldRenderSuggestions(0, true)).toBe(false);
    expect(wrapper.vm.shouldRenderSuggestions(1, true)).toBe(false);
    expect(wrapper.vm.shouldRenderSuggestions(1, false)).toBe(true);
  });

  it('should render geocoder suggestion', () => {
    const wrapper = mountWithOptions(store);
    const suggestion = wrapper.vm.renderSuggestion({
      name: 'geocoder',
      item: {
        description: 'mock description',
      },
    });
    expect(suggestion.children[0].context.$el.outerHTML).toMatchSnapshot();
  });

  it('should render suggestion', () => {
    const wrapper = mountWithOptions(store);
    const suggestion = wrapper.vm.renderSuggestion({
      name: 'notgeocoder',
      item: {
        description: 'mock description',
        ...MockWorksites[0],
      },
    });
    expect(suggestion.children[0].context.$el.outerHTML).toMatchSnapshot();
  });

  it('should render correctly and match snapshot', () => {
    const wrapper = mountWithOptions(store);
    expect(wrapper.element).toMatchSnapshot();
  });
});
