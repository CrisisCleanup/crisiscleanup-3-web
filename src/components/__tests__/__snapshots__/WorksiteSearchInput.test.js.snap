// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorksiteSearchInput should render correctly and match snapshot 1`] = `
<div
  class="flex items-center justify-start w-full autocomplete"
>
  <vue-autosuggest-stub
    class="flex-grow relative text-base font-light base"
    get-suggestion-value="function () { [native code] }"
    input-props="[object Object]"
    render-suggestion="function () { [native code] }"
    section-configs="[object Object]"
    should-render-suggestions="function () { [native code] }"
    suggestions="[object Object]"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`WorksiteSearchInput should render geocoder suggestion 1`] = `
VueComponent {
  "$attrs": Object {},
  "$children": Array [],
  "$createElement": [Function],
  "$listeners": Object {},
  "$options": Object {
    "_Ctor": Object {
      "0": [Function],
    },
    "_base": [Function],
    "beforeCreate": Array [
      [Function],
      [Function],
    ],
    "components": Object {},
    "directives": Object {},
    "filters": Object {},
    "mounted": Array [
      [Function],
    ],
    "updated": Array [
      [Function],
    ],
  },
  "$parent": undefined,
  "$refs": Object {},
  "$root": [Circular],
  "$scopedSlots": Object {},
  "$slots": Object {},
  "$vnode": undefined,
  "_c": [Function],
  "_data": Object {},
  "_directInactive": false,
  "_events": Object {},
  "_hasHookEvent": false,
  "_inactive": null,
  "_isBeingDestroyed": false,
  "_isDestroyed": false,
  "_isMounted": false,
  "_isVue": true,
  "_renderProxy": VueComponent {
    "$attrs": Object {},
    "$children": Array [],
    "$createElement": [Function],
    "$listeners": Object {},
    "$options": Object {
      "_Ctor": Object {
        "0": [Function],
      },
      "_base": [Function],
      "beforeCreate": Array [
        [Function],
        [Function],
      ],
      "components": Object {},
      "directives": Object {},
      "filters": Object {},
      "mounted": Array [
        [Function],
      ],
      "updated": Array [
        [Function],
      ],
    },
    "$parent": undefined,
    "$refs": Object {},
    "$root": [Circular],
    "$scopedSlots": Object {},
    "$slots": Object {},
    "$vnode": undefined,
    "_c": [Function],
    "_data": Object {},
    "_directInactive": false,
    "_events": Object {},
    "_hasHookEvent": false,
    "_inactive": null,
    "_isBeingDestroyed": false,
    "_isDestroyed": false,
    "_isMounted": false,
    "_isVue": true,
    "_renderProxy": [Circular],
    "_self": [Circular],
    "_staticTrees": null,
    "_uid": 20,
    "_vnode": null,
    "_watcher": null,
    "_watchers": Array [],
  },
  "_self": [Circular],
  "_staticTrees": null,
  "_uid": 20,
  "_vnode": null,
  "_watcher": null,
  "_watchers": Array [],
}
`;

exports[`WorksiteSearchInput should render suggestion 1`] = `
VueComponent {
  "$attrs": Object {},
  "$children": Array [],
  "$createElement": [Function],
  "$listeners": Object {},
  "$options": Object {
    "_Ctor": Object {
      "0": [Function],
    },
    "_base": [Function],
    "beforeCreate": Array [
      [Function],
      [Function],
    ],
    "components": Object {},
    "directives": Object {},
    "filters": Object {},
    "mounted": Array [
      [Function],
    ],
    "updated": Array [
      [Function],
    ],
  },
  "$parent": undefined,
  "$refs": Object {},
  "$root": [Circular],
  "$scopedSlots": Object {},
  "$slots": Object {},
  "$vnode": undefined,
  "_c": [Function],
  "_data": Object {},
  "_directInactive": false,
  "_events": Object {},
  "_hasHookEvent": false,
  "_inactive": null,
  "_isBeingDestroyed": false,
  "_isDestroyed": false,
  "_isMounted": false,
  "_isVue": true,
  "_renderProxy": VueComponent {
    "$attrs": Object {},
    "$children": Array [],
    "$createElement": [Function],
    "$listeners": Object {},
    "$options": Object {
      "_Ctor": Object {
        "0": [Function],
      },
      "_base": [Function],
      "beforeCreate": Array [
        [Function],
        [Function],
      ],
      "components": Object {},
      "directives": Object {},
      "filters": Object {},
      "mounted": Array [
        [Function],
      ],
      "updated": Array [
        [Function],
      ],
    },
    "$parent": undefined,
    "$refs": Object {},
    "$root": [Circular],
    "$scopedSlots": Object {},
    "$slots": Object {},
    "$vnode": undefined,
    "_c": [Function],
    "_data": Object {},
    "_directInactive": false,
    "_events": Object {},
    "_hasHookEvent": false,
    "_inactive": null,
    "_isBeingDestroyed": false,
    "_isDestroyed": false,
    "_isMounted": false,
    "_isVue": true,
    "_renderProxy": [Circular],
    "_self": [Circular],
    "_staticTrees": null,
    "_uid": 20,
    "_vnode": null,
    "_watcher": null,
    "_watchers": Array [],
  },
  "_self": [Circular],
  "_staticTrees": null,
  "_uid": 20,
  "_vnode": null,
  "_watcher": null,
  "_watchers": Array [],
}
`;

exports[`WorksiteSearchInput should retrieve current user 1`] = `
Object {
  "first_name": "First",
  "id": 1,
  "last_name": "Last",
  "organization": Object {
    "id": 1,
    "name": "Test Organization",
  },
}
`;
